pallets/astar-xcm-benchmarks/src/generic/benchmarking.rs

```diff
@@ -0,0 +1,151 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+use super::{Pallet as AstarBenchmarks, *};
+use crate::WrappedBenchmark;
+use frame_benchmarking::v2::*;
+use frame_support::dispatch::Weight;
+use pallet_xcm_benchmarks::{generic::Pallet as PalletXcmBenchmarks, new_executor, XcmCallOf};
+use sp_std::vec;
+use xcm::latest::prelude::*;
+
+#[benchmarks]
+mod benchmarks {
+    use super::*;
+
+    /// We need re-write buy_execution benchmark becuase our runtime
+    /// needs 1 additional DB read (XcAssetConfig) for fetching unit per sec
+    /// for a fungible asset. The upstream benchmark use native assets thus
+    /// won't accout for it.
```

Comment:
One general question about this approach - how viable is to reuse code from the official repo if we have to account for different actions for some calls?

This can change in the future, upstream, and it means we should be re-checking all of the benchmarks each time we do an uplift?

It would seem easier to just have all the benchmarks in our repo since that way we have full control over it. If I understood the current approach correctly, it's highly likely we miss something in the future.

Reply:
The upstream benchmarks are quite generic themselves, but not generic enough maybe because not many parachains are using them. 
The benchmarks we had to re-write were only due the use of hardcoded native asset inside some benchmarks.
- **buy_execution**: hardcoded native asset used for buying execution
- **transfer_***: does not take ED into account, they use native asset (pallet_balances) which defaults to allow death unlike fungibles (`pallet_assets`).
- **expect_pallet**: hardcoded pallet index

Other than that all benchmarks were configurable through the pallet's Config. As long as we we used same release upstream pallet (`XcmConfig` is not changed) it should be fine.

For the hardcoded cases, we can make a upstream PR to make it more generic for other parachains to use as well.

I'm fine either ways, if we copy-paste all benchmarks then we need to keep them updated during uplift if there are major changes like `precompile-utils` and if we don't copy-paste then we still need to check them during uplifts

Reply:
Thanks for the explanation!

Let's keep it like this, as you suggested.
For the upstream PR, let's start with an issue first perhaps?